File: AbMath/Calculator/Simplifications/Sqrt.cs
using System;
using System.Collections.Generic;
using System.Text;
using AbMath.Calculator.Operators;

namespace AbMath.Calculator.Simplifications
{
    public static class Sqrt
    {
        public static bool SqrtNegativeNumbersRunnable(RPN.Node node)
        {
            return node.IsSqrt() && node[0].IsLessThanNumber(0);
        }

        public static RPN.Node SqrtNegativeNumbers(RPN.Node node)
        {
            return new RPN.Node(double.NaN);
        }

        public static bool SqrtToFuncRunnable(RPN.Node node)
        {
            return node.IsExponent() && node[0].IsNumber(2) && node[1].IsSqrt();
        }

        public static RPN.Node SqrtToFunc(RPN.Node node)
        {
            return node[1, 0];
        }

        public static bool SqrtToAbsRunnable(RPN.Node node)
        {
            return node.IsSqrt() && node[0].IsExponent() && node[0,0].IsNumber(2);
        }

        public static RPN.Node SqrtToAbs(RPN.Node node)
        {
            return new RPN.Node(new[] { node[0,1] }, new RPN.Token("abs", 1, RPN.Type.Function));
        }

        public static bool SqrtPowerFourRunnable(RPN.Node node)
        {
            return node.IsSqrt() && node[0].IsExponent() &&
                   node[0,0].IsNumber() &&
                   node[0,0].GetNumber() % 4 == 0;
        }

        public static RPN.Node SqrtPowerFour(RPN.Node node)
        {
            return new Pow(node[0, 1], new RPN.Node(node[0,0].GetNumber() / 2));
        }
    }
}

File: rust/abmath/src/rules/sqrt.rs
use crate::ast::Node;
use crate::rules::Rule;
use crate::tokenizer::{MathFunction, MathOperator};

pub struct SqrtNegativeNumbersRule;
impl Rule for SqrtNegativeNumbersRule {
    fn name(&self) -> &'static str {
        "SqrtNegativeNumbers"
    }

    fn apply(&self, node: &Node) -> Option<Node> {
        // sqrt(-x) -> NaN (unless handling complex numbers)
        if let Node::Function(MathFunction::Sqrt, args) = node {
            if let Some(arg0) = args.first() {
                if arg0.less_than_number(0.0) {
                    return Some(Node::Number(f64::NAN));
                }
            }
        }
        None
    }
}

pub struct SqrtToFuncRule;
impl Rule for SqrtToFuncRule {
    fn name(&self) -> &'static str {
        "SqrtToFunc"
    }

    fn apply(&self, node: &Node) -> Option<Node> {
        // sqrt(x)^2 -> x
        if let Node::BinaryOp(MathOperator::Pow, left_base, right_exp) = node {
            if right_exp.is_number(2.0) {
                if let Node::Function(MathFunction::Sqrt, sqrt_args) = &**left_base {
                    if let Some(inner) = sqrt_args.first() {
                        return Some(inner.clone());
                    }
                }
            }
        }
        None
    }
}

pub struct SqrtToAbsRule;
impl Rule for SqrtToAbsRule {
    fn name(&self) -> &'static str {
        "SqrtToAbs"
    }

    fn apply(&self, node: &Node) -> Option<Node> {
        // sqrt(x^2) -> abs(x)
        if let Node::Function(MathFunction::Sqrt, args) = node {
            if let Some(arg0) = args.first() {
                if let Node::BinaryOp(MathOperator::Pow, pow_base, pow_exp) = arg0 {
                    if pow_exp.is_number(2.0) {
                        return Some(Node::Function(MathFunction::Abs, vec![*pow_base.clone()]));
                    }
                }
            }
        }
        None
    }
}

pub struct SqrtPowerFourRule;
impl Rule for SqrtPowerFourRule {
    fn name(&self) -> &'static str {
        "SqrtPowerFour"
    }

    fn apply(&self, node: &Node) -> Option<Node> {
        // sqrt(x^y) where y % 4 == 0 -> x^(y/2)
        // Let's actually just do y % 2 == 0 more generally if it's even unless the C# code strictly requires % 4?
        // Ah, the C# code says `node[0,0].GetNumber() % 4 == 0` and returns `Pow(base, num/2)`.
        // We'll stick strictly to `% 4 == 0` to identically match `SqrtPowerFourRunnable`.
        if let Node::Function(MathFunction::Sqrt, args) = node {
            if let Some(arg0) = args.first() {
                if let Node::BinaryOp(MathOperator::Pow, pow_base, pow_exp) = arg0 {
                    if let Some(val) = pow_exp.get_number() {
                        if val.fract() == 0.0 && val % 4.0 == 0.0 {
                            return Some(Node::BinaryOp(
                                MathOperator::Pow,
                                pow_base.clone(),
                                Box::new(Node::Number(val / 2.0)),
                            ));
                        }
                    }
                }
            }
        }
        None
    }
}
